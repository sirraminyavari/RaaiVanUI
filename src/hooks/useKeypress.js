import { useState, useEffect } from 'react';

/**
 * This hook makes it easy to detect when the user is pressing a specific key on their keyboard.
 * @hook
 * @param {string} targetKey -The key name that user has pressed.
 * @returns {boolean} The result as boolean which indicates if key has been pressed or not.
 */
const useKeyPress = (targetKey) => {
  //! State for keeping track of whether key is pressed
  const [keyPressed, setKeyPressed] = useState(false);

  //! If pressed key is our target key then set to true
  const downHandler = ({ key }) => {
    if (key === targetKey) {
      setKeyPressed(true);
    }
  };

  //! If released key is our target key then set to false
  const upHandler = ({ key }) => {
    if (key === targetKey) {
      setKeyPressed(false);
    }
  };

  //! Add event listeners
  useEffect(() => {
    window.addEventListener('keydown', downHandler);
    window.addEventListener('keyup', upHandler);
    //! Remove event listeners on cleanup
    return () => {
      window.removeEventListener('keydown', downHandler);
      window.removeEventListener('keyup', upHandler);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return keyPressed;
};

export default useKeyPress;
