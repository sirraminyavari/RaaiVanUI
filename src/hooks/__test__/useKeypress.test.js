import { render, cleanup, fireEvent } from '@testing-library/react';

import useKeypress from '../useKeypress';

describe('useKeypress Test', () => {
  afterEach(cleanup);
  let wrapper;

  const renderHook = (hook, targetKey, testKey, toBe) => {
    let result;

    const HookWrapper = () => {
      result = hook(testKey);
      return null;
    };

    wrapper = render(<HookWrapper />);
    fireEvent.keyDown(window, { key: targetKey, code: targetKey });

    expect(result).toBe(toBe);

    return result;
  };

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('works and catches correct key press', () => {
    const result = renderHook(useKeypress, 'Enter', 'Enter', true);

    console.log(result);
    wrapper.debug();
  });

  it('works and does not catche incorrect key press', () => {
    const result = renderHook(useKeypress, 'Escape', 'Enter', false);

    console.log(result);
    wrapper.debug();
  });
});
