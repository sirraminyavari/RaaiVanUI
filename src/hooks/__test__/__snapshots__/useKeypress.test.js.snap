// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useKeypress Test matches snapshot 1`] = `undefined`;
